File: src/main/java/leetcode/JumpGame.java
package leetcode;

/**
 * Given an array of non-negative integers nums, you are initially positioned at
 * the first index of the array.
 *
 * Each element in the array represents your maximum jump length at that
 * position.
 *
 * Determine if you are able to reach the last index.
 */
public class JumpGame {

    /**
     * Determine if you can reach the last index.
     *
     * The time complexity is O(2^N) as you make 2 decisions N times.
     *
     * The space complexity is O(N) for the recursive stack frames.
     *
     * @param nums The input array.
     * @return True if you can and false if you cannot.
     */
    public boolean canJump(final int[] nums) {
        return canJumpFromPosition(nums, 0);
    }

    /**
     * Helper method for recursion.
     *
     * @param nums     The input array.
     * @param position The current position.
     * @return True if you can to the next position and false if you cannot.
     */
    private boolean canJumpFromPosition(final int[] nums, final int position) {

        // we have reached the end
        if (position == nums.length - 1) {
            return true;
        }

        int furthestJump = nums[position];

        // do all of the jumps
        for (var i = 1; i <= furthestJump; i++) {
            if (canJumpFromPosition(nums, position + i)) {
                return true;
            }
        }

        return false;
    }

}

File: python/src/jump_game.py
from typing import List


class Solution:
    """
    You are given an integer array nums. You are initially positioned at the array's first index, and each element in
    the array represents your maximum jump length at that position.

    Return true if you can reach the last index, or false otherwise.
    """

    def __init__(self):
        """
        Initialise the solution.
        """
        self.memo = None
        self.res = None

    def canJump_naive(self, nums: List[int]) -> bool:
        """
        Return true if you can reach the last index, or false otherwise.

        :param nums: The integer array.
        :type nums: List[Int].
        :return: True if you can jump to the end of the array, and False otherwise.
        :rtype: Bool.

        The time complexity is O(N^N). Results in Time Limit Exceeded.

        The space complexity is O(N) for the recursive stack. Results in Time Limit Exceeded.
        """
        self.res = False
        self.dfs_naive(0, nums)
        return self.res

    def canJump_memo(self, nums: List[int]) -> bool:
        """
        Return true if you can reach the last index, or false otherwise.

        :param nums: The integer array.
        :type nums: List[Int].
        :return: True if you can jump to the end of the array, and False otherwise.
        :rtype: Bool.

        The time complexity is O(2^N). Results in Time Limit Exceeded.

        The space complexity is O(N) for the recursive stack and memoization dict. Results in Time Limit Exceeded.
        """
        self.memo = {}
        self.dfs_memo(0, nums)
        return self.dfs_memo(0, nums)

    def canJump_dfs(self, nums: List[int]) -> bool:
        """
        Return true if you can reach the last index, or false otherwise.

        :param nums: The integer array.
        :type nums: List[Int].
        :return: True if you can jump to the end of the array, and False otherwise.
        :rtype: Bool.

        The time complexity is O(N). Faster than 97.00% of solutions.

        The space complexity is O(1) as we don't use any additional data structures. Less memory than 49.40% of
        solutions.
        """
        goal = len(nums) - 1
        for i in range(len(nums) - 1, -1, -1):
            # if we can get to the end from here then this is the new goal
            if i + nums[i] >= goal:
                goal = i

        # if goal is 0 we can get to the end from the start!
        return True if goal == 0 else False

    def dfs_memo(self, i: int, nums: List[int]):
        """
        A helper method to perform a Depth-First Search (DFS).

        :param i: The current index.
        :type i: Int.
        :param nums: The integer array.
        :type nums: List[Int].
        :return: NoneType.
        :rtype: NoneType.
        """
        if i >= len(nums):
            return False
        elif i == len(nums) - 1:
            return True
        elif i in self.memo:
            return self.memo[i]
        else:
            self.memo[i] = False
            jumps = nums[i]
            while jumps > 0:
                if self.dfs_memo(i + jumps, nums):
                    self.memo[i] = True
                    return self.memo[i]
                jumps -= 1
            return self.memo[i]

    def dfs_naive(self, i: int, nums: List[int]):
        """
        A helper method to perform a Depth-First Search (DFS).

        :param i: The current index.
        :type i: Int.
        :param nums: The integer array.
        :type nums: List[Int].
        :return: NoneType.
        :rtype: NoneType.
        """
        if i >= len(nums) - 1:
            self.res = True
            return
        jumps = nums[i]
        while jumps > 0:
            self.dfs_naive(i + jumps, nums)
            jumps -= 1
